esrally/mechanic/launcher.py

```diff
@@ -122,83 +127,32 @@ def stop(self, c):
-                lines_to_log -= 1
-
-            # no need to check as soon as we have detected node startup
-            if not self.startup_event.isSet():
-                if line.find("Initialization Failed") != -1 or line.find("A fatal exception has occurred") != -1:
-                    self.logger.error("[%s] encountered initialization errors.", self.node_name)
-                    # wait a moment to ensure the process has terminated before we signal that we detected a (failed) startup.
-                    self.await_termination(self.server)
-                    self.startup_event.set()
-                if line.endswith("started") and not self.startup_event.isSet():
-                    self.startup_event.set()
-                    self.logger.info("[%s] has successfully started.", self.node_name)
-
-    def await_termination(self, server, timeout=5):
-        # wait a moment to ensure the process has terminated
-        wait = timeout
-        while not server.returncode or wait == 0:
-            time.sleep(0.1)
-            server.poll()
-            wait -= 1
-
-
-def _start(process, node_name):
-    log = logging.getLogger(__name__)
-    startup_event = threading.Event()
-    watcher = StartupWatcher(node_name, process, startup_event)
-    t = threading.Thread(target=watcher.watch)
-    t.setDaemon(True)
-    t.start()
-    if startup_event.wait(timeout=InProcessLauncher.PROCESS_WAIT_TIMEOUT_SECONDS):
-        process.poll()
-        # has the process terminated?
-        if process.returncode:
-            msg = "Node [%s] has terminated with exit code [%s]." % (node_name, str(process.returncode))
-            log.error(msg)
-            raise exceptions.LaunchError(msg)
-        else:
-            log.info("Started node [%s] with PID [%s].", node_name, process.pid)
-            return process
-    else:
-        msg = "Could not start node [%s] within timeout period of [%s] seconds." % (
-            node_name, InProcessLauncher.PROCESS_WAIT_TIMEOUT_SECONDS)
-        # check if the process has terminated already
-        process.poll()
-        if process.returncode:
-            msg += " The process has already terminated with exit code [%s]." % str(process.returncode)
-        else:
-            msg += " The process seems to be still running with PID [%s]." % process.pid
-        log.error(msg)
-        raise exceptions.LaunchError(msg)
+    output = subprocess.check_output(args=shlex.split(compose_ps_cmd))
+    return output.decode("utf-8").rstrip()
+
+
+def _wait_for_healthy_running_container(container_id, timeout=60):
+    import docker
+    cli = docker.api.APIClient()
+
+    endtime = _time() + timeout
+    while _time() < endtime:
+        containers = cli.containers(quiet=True, all=True, filters={"id": container_id,
+                                                                   "health": "healthy",
+                                                                   "status": "running"})
```

Comment:
So if I understand correctly this is the only place where we use the imported docker library.

There is no question that this is cleaner from a code PoV, however, this adds a dependency that apart from needing to be listed (incl. its license, apache2) in `setup.py` it pulls a number of transitive dependencies (see below.

Ideally we want to keep the number of dependencies to the minimum and some of the dependent libraries are also used by Rally, hence have the potential for version conflicts in the future.

Given all this, does it make sense to just rely on explicit docker cli invocation here and not rely on the library? Did you hit any issue that I am overlooking that made you steer towards the library?


```
$ pip3 install docker
Collecting docker
  Using cached https://files.pythonhosted.org/packages/91/93/310fe092039f6b0759a1f8524e9e2c56f8012804fa2a8da4e4289bb74d7c/docker-4.0.1-py2.py3-none-any.whl
Requirement already satisfied: six>=1.4.0 in ./.venv/lib/python3.6/site-packages (from docker) (1.12.0)
Requirement already satisfied: requests!=2.18.0,>=2.14.2 in ./.venv/lib/python3.6/site-packages (from docker) (2.21.0)
Collecting websocket-client>=0.32.0 (from docker)
  Using cached https://files.pythonhosted.org/packages/29/19/44753eab1fdb50770ac69605527e8859468f3c0fd7dc5a76dd9c4dbd7906/websocket_client-0.56.0-py2.py3-none-any.whl
Requirement already satisfied: idna<2.9,>=2.5 in ./.venv/lib/python3.6/site-packages (from requests!=2.18.0,>=2.14.2->docker) (2.8)
Requirement already satisfied: urllib3<1.25,>=1.21.1 in ./.venv/lib/python3.6/site-packages (from requests!=2.18.0,>=2.14.2->docker) (1.22)
Requirement already satisfied: certifi>=2017.4.17 in ./.venv/lib/python3.6/site-packages (from requests!=2.18.0,>=2.14.2->docker) (2018.11.29)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in ./.venv/lib/python3.6/site-packages (from requests!=2.18.0,>=2.14.2->docker) (3.0.4)
Installing collected packages: websocket-client, docker
Successfully installed docker-4.0.1 websocket-client-0.56.0
```

Reply:
I only used the SDK due to a general preference for APIs over CLIs in my code, but it does not make much sense here.